File: Training/Trainer.cs
using ACommerce.MagneticLM.Graph;

namespace ACommerce.MagneticLM.Training;

public class Trainer
{
    private readonly WordGraph _graph;
    public int SentencesTrained { get; private set; }

    public Trainer(WordGraph graph) => _graph = graph;

    public void TrainSentence(string sentence)
    {
        var words = Tokenize(sentence);
        if (words.Length < 2) return;

        // === 1. تسجيل n-grams بكل الأعماق ===
        for (int i = 0; i < words.Length; i++)
        {
            var node = _graph.GetOrCreateNode(words[i]);
            node.Frequency++;
            _graph.TotalTokens++;

            if (i < words.Length - 1)
            {
                var nextWord = words[i + 1];

                // unigram→bigram→trigram→...→MaxNgramOrder
                for (int order = 1; order <= _graph.MaxNgramOrder && order <= i + 1; order++)
                {
                    var context = words[(i + 1 - order)..(i + 1)];
                    _graph.AddNgram(context, nextWord);
                }
            }
        }

        // === 2. اكتشاف معنوي + مكافأة/عقوبة (Reward/Penalty) ===
        // كل جملة تدريب هي أيضاً عملية استدلال:
        //   - نتوقع الكلمة التالية
        //   - إذا نجح التوقع → مكافأة (تقوية الوزن المعنوي)
        //   - إذا فشل → عقوبة (إضعاف الأوزان الخاطئة)
        for (int i = 0; i < words.Length; i++)
        {
            // 2a. أوزان معنوية من نافذة ±2
            for (int d = -2; d <= 2; d++)
            {
                if (d == 0) continue;
                int j = i + d;
                if (j < 0 || j >= words.Length) continue;
                var weight = Math.Abs(d) <= 1 ? 1.0 : _graph.TransitiveDecay;
                _graph.StrengthSemanticEdge(words[i], words[j], weight * 0.1);
            }

            // 2b. مكافأة/عقوبة: هل الطبقة المعنوية تتوقع الكلمة التالية بشكل صحيح؟
            if (i < words.Length - 1)
            {
                var actual = words[i + 1];
                var neighbors = _graph.GetSemanticNeighbors(words[i]).Take(20).ToList();

                foreach (var (neighbor, w) in neighbors)
                {
                    if (neighbor == actual)
                    {
                        // مكافأة: التوقع المعنوي كان صحيحاً!
                        _graph.StrengthSemanticEdge(words[i], actual, 0.05);
                    }
                    else if (w > 0.5)
                    {
                        // عقوبة: وزن معنوي عالي لكلمة لم تظهر فعلاً
                        _graph.StrengthSemanticEdge(words[i], neighbor, -0.02);
                    }
                }
            }
        }

        // الأوزان الضعيفة لا تُحذف - تُتجاهل عند الاستعلام فقط
        // الأوزان السالبة القوية = معرفة طرد مفيدة تبقى

        // === 3. إعادة التقييم العابر ===
        SentencesTrained++;
        if (SentencesTrained % 200 == 0)
            PropagateTransitiveRelations(words);
        if (SentencesTrained % 1000 == 0)
            Console.Write($"\r  Training: {SentencesTrained:N0} sentences...");
    }

    private void PropagateTransitiveRelations(string[] words)
    {
        foreach (var word in words)
        {
            var neighbors = _graph.GetSemanticNeighbors(word)
                .OrderByDescending(n => n.Weight).Take(8).ToList();

            foreach (var (neighbor, w1) in neighbors)
            {
                if (w1 < _graph.SemanticThreshold) continue;

                foreach (var (transitive, w2) in _graph.GetSemanticNeighbors(neighbor)
                    .OrderByDescending(n => n.Weight).Take(8))
                {
                    if (transitive == word || w2 < _graph.SemanticThreshold) continue;

                    var tw = w1 * w2 * _graph.TransitiveDecay * 0.01;
                    if (tw > 0.001)
                        _graph.StrengthSemanticEdge(word, transitive, tw);
                }
            }
        }
    }

    // لا نحذف أوزاناً - السالبة القوية = طرد مفيد، الضعيفة تُتجاهل عند الاستعلام

    public void TrainBatch(IEnumerable<string> sentences)
    {
        foreach (var s in sentences) TrainSentence(s);
        if (SentencesTrained >= 1000) Console.WriteLine();
    }

    public static string[] Tokenize(string sentence)
    {
        return sentence
            .Replace(".", " ").Replace("،", " ").Replace(",", " ")
            .Replace("!", " ").Replace("؟", " ").Replace("?", " ")
            .Replace("\"", " ").Replace("(", " ").Replace(")", " ")
            .Split(' ', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            .Where(w => w.Length > 0)
            .ToArray();
    }
}

File: colab/trainer.py
"""
MagneticLM Trainer - OPTIMIZED
Training as accounting entries with batch processing.
"""

from graph import WordGraph
from typing import List
import re

# Pre-compiled regex for fast tokenization
_SPLIT_PATTERN = re.compile(r'[.,;!?؟،"()\[\]{}]+')


def tokenize(sentence: str) -> List[str]:
    """Optimized tokenizer."""
    cleaned = _SPLIT_PATTERN.sub(' ', sentence.lower())
    return [w for w in cleaned.split() if len(w) > 0]


class Trainer:
    def __init__(self, graph: WordGraph):
        self.graph = graph
        self.sentences_trained = 0

    def train_sentence(self, sentence: str):
        words = tokenize(sentence)
        if len(words) < 2:
            return

        g = self.graph  # local reference for speed

        # === 1. Register n-grams ===
        for i in range(len(words)):
            node = g.get_or_create(words[i])
            node.freq += 1
            g.total_tokens += 1

            if i < len(words) - 1:
                next_word = words[i + 1]
                for order in range(1, min(g.max_order + 1, i + 2)):
                    context = tuple(words[i + 1 - order:i + 1])
                    g.add_ngram(context, next_word)

        # === 2. Semantic (depth ±2) + Reward/Penalty ===
        semantic_add = g.add_semantic  # local ref
        for i in range(len(words)):
            w = words[i]
            for d in (-2, -1, 1, 2):
                j = i + d
                if 0 <= j < len(words):
                    semantic_add(w, words[j], (1.0 if abs(d) <= 1 else 0.5) * 0.1)

            if i < len(words) - 1:
                actual = words[i + 1]
                neighbors = list(g.semantic.get(w, {}).items())[:20]
                for neighbor, wt in neighbors:
                    if neighbor == actual:
                        semantic_add(w, actual, 0.05)
                    elif wt > 0.5:
                        semantic_add(w, neighbor, -0.02)

        # === 3. Transitive propagation ===
        self.sentences_trained += 1
        if self.sentences_trained % 200 == 0:
            self._propagate(words)
        if self.sentences_trained % 1000 == 0:
            print(f"\r  Training: {self.sentences_trained:,} sentences...", end="", flush=True)

    def _propagate(self, words: List[str]):
        threshold = self.graph.semantic_threshold
        for word in words:
            neighbors = sorted(
                ((n, w) for n, w in self.graph.semantic.get(word, {}).items() if abs(w) >= threshold),
                key=lambda x: -abs(x[1])
            )[:8]
            for neighbor, w1 in neighbors:
                for trans, w2 in sorted(
                    ((n, w) for n, w in self.graph.semantic.get(neighbor, {}).items() if abs(w) >= threshold),
                    key=lambda x: -abs(x[1])
                )[:8]:
                    if trans == word: continue
                    tw = w1 * w2 * 0.5 * 0.01
                    if tw > 0.001:
                        self.graph.add_semantic(word, trans, tw)

    def train_batch(self, sentences):
        for s in sentences:
            self.train_sentence(s)
        if self.sentences_trained >= 1000:
            print()
